Improved solve_univariate_inequality function

arihantparsoya opened the pull request:
Previously only the inequalities which have roots could be solved and as a result of which inequalities such as solveset(x + sqrt(x**2-5)>0,x,S.Reals) was giving wrong result (issue #11064):
#### previously:

`In [4]: solveset(x+sqrt(x**2-5)>0,x,S.Reals)`
`Out[4]: ∅` 
#### this PR give output:

`In [4]: solveset(x+sqrt(x**2-5)>0,x,S.Reals)`
`Out[4]: (sqrt(5), oo)` 
- Added function _inequalities_without_roots which evaluates the sub expression of the original equation and the final result is the intersection of solutions of the individual subexpression
- it is important to note that the function is called only when there are no roots for the original expression.

arihantparsoya committed:
Merge remote-tracking branch 'sympy/master'

arihantparsoya committed:
Merge remote-tracking branch 'sympy/master'

arihantparsoya committed:
Merge remote-tracking branch 'sympy/master'

arihantparsoya committed:
added functionality to solveset

Previously only the inequalities which have roots could be solved and
as a result of which inequalities such as solveset(x +
sqrt(x**2-5)>0,x,S.Reals) was giving wrong result:

- Added function _inequalities_without_roots which evaluates the sub
expression of the original equation and the final result is the
intersection of solutions of the individual subexpression

- it is important to note that the function is called only when there
are no roots for the original expression.

arihantparsoya committed:
added test

this is fix for the issue #11064:

- added test

parsoyaarihant commented:
@Shekharrajak , @aktech , can you review this?

Shekharrajak commented:
Please give proper title.[wiki/Development-workflow#writing-pull-request-title-and-description](https://github.com/sympy/sympy/wiki/Development-workflow#writing-pull-request-title-and-description).
Add references in the description section.

kshitij10496 commented on sympy/solvers/inequalities.py:
```diff
@@ -481,11 +482,25 @@ def valid(x):
+    e = expr.gts - expr.lts
+    if e.is_Add:
+        for subexpr in e.args:
+            n_sol_sets.append(solve_univariate_inequality(subexpr > 0, gen, relational = False))
```
I have a concern with this.
How do you know the kind of inequality we are looking for ?

To be precise, what was the reason behind generalising `subexpr > 0` (independent of the type of inequality) ?

parsoyaarihant replied:
- the expression `e = expr.gts - expr.lts` makes sure that `subexpr>0`.

- Also `e.is_Add = True` , (line 495) makes sure that `e` subfunctions are being added (not multiplied or divided)

smichr replied:
When supplying keywords, omit space around `=`: `relational=False`

parsoyaarihant replied:
Thanks for pointing out :) .

parsoyaarihant commented:
@Shekharrajak , thanks for pointing out. I will make sure this never happens again :) .

I have made the required changes.

arihantparsoya committed:
Typo error

smichr commented:
I have the feeling that this functionality should be added to a function named `real_conditions` which returns the domain under which a function will be real. e.g. `solveset(1/sqrt(x)>0)` could give `x > 0`. I don't think the current function would give that result.

parsoyaarihant commented:
the equaiton `solveset(1/sqrt(x)>0)` can already be solved by the funciton solve_univariate_inequality since the equation `1/sqrt(x)` has a singularity at `x = 0`.

smichr commented:
What about `solveset(2 + sqrt(x - 4) > 0)`? There are no zeros to the function, the real domain is `x>=4`; sampling at `x=4` gives a value of 2 (which is positive) so this inequality is always positive for `x >= 4`.

arihantparsoya committed:
Changed the name of function

- changed the function name to _real_conditions

- set the variable check=True in line 420.

- equations like ‘ 2 + sqrt(x - 4) ‘ does not have any solution but
because of check=False wrong solution was being returned.

- Now correct result is being evaluated for the equation `2 + sqrt(x -
4)>0`

parsoyaarihant commented on sympy/solvers/inequalities.py:
```diff
@@ -417,7 +417,7 @@ def solve_univariate_inequality(expr, gen, relational=True):
             singularities = solve(d, gen, check=False)
         else:
-            solns = solve(e, gen, check=False)
+            solns = solve(e, gen, check=True)
```
@smichr , due to check=False wrong solution to the equation `(2 + sqrt(x - 4)` being returned and the function _real_conditions wasnt being called. Now, correct solution is being returned i.e. `(4, oo)`

parsoyaarihant commented:
@smichr , I commented on the line 420 of solveset.py.

arihantparsoya committed:
added test

added test for the equation

solveset(eq > 0, x, S.Reals)

parsoyaarihant commented:
There is a problem with the function solve when evaluating `solve(sqrt(x) + 1/root(x, 3) - 1 , test=True)` , its taking too long to test the solution. Hence, [test_slow_general_univariate()](https://github.com/parsoyaarihant/sympy/blob/solveset-bug-%2311064/sympy/solvers/tests/test_inequalities.py#L302) is giving wrong result. So far, I couldnt figure out how to make function `solve_univariate_inequality` robust enough to satisfy both the equations : 
 - `solveset(2 + sqrt(x - 4) > 0, x, S.Reals)`
 - `solve(sqrt(x) + 1/root(x, 3) > 1)`

@Shekharrajak , any ideas on this one?

arihantparsoya committed:
reverted back changes

by using Check = True,

solve(sqrt(x) + 1/root(x, 3) > 1), could not be solved and by using
Check = False the equation solveset(2 + sqrt(x - 4) > 0, x, S.Reals)
give wrong answer

Shekharrajak commented:
@parsoyaarihant , It seems your attempt is going on right way.  But you need to think about this case

```
In [ ]: solveset(x**2 + sqrt(x**2-5) >0,x,S.Reals)
Out[ ]: ∅
```
Ans is ` [sqrt(5), infinity) union (-infinity, -sqrt(5)] `

Shekharrajak commented on sympy/solvers/inequalities.py:
```diff
@@ -481,11 +482,24 @@ def valid(x):
+    e = expr.gts - expr.lts
+    if e.is_Add:
+        for subexpr in e.args:
+            n_sol_sets.append(solve_univariate_inequality(subexpr > 0, gen, relational=False))
```
Hint : You need to check the `domain` of `subexpr`. It is not every time `(0, oo)` .

parsoyaarihant replied:
since `e = expr.gts - expr.lts` , it makes sure that `subexpr > 0` since `expr.gts` give the equation on the `greater than side` : [reference](https://github.com/sympy/sympy/blob/master/sympy/core/relational.py#L451)

Shekharrajak replied:
What about `less than` ? 
```
In [2]: solveset(x + sqrt(x**2-5) <0,x,S.Reals)
Out[2]: ∅
```

parsoyaarihant replied:
For:
` In [2]: solveset(x + sqrt(x**2-5) <0,x,S.Reals)`
` Out[2]: ∅ `
the output is correct since if u take two equations:
- `sqrt(x**2-5) < 0` --(1)
- `x  < 0` --(2)
- equation (1) has no solution and hence (1) + (2) has no solution.

Shekharrajak replied:
Put `x` = `- sqrt(6) `

Shekharrajak replied:
>  the output is correct since if you take two equations:
> 
> sqrt(x**2-5) < 0 --(1)
> x < 0 --(2)

Where did you learn this ?

Shekharrajak replied:
Actually we do 
` sqrt(x**2-5) < -x ` since `sqrt(x**2 - 5) > 0` so ` -x  > 0`.
Please use online solvers, that may help.

parsoyaarihant replied:
Sorry about not checking the solution properly. But the algorithm I implemented in `_real_function()` doesn't account for this equation. I will work on it.

parsoyaarihant replied:
@Shekharrajak , I made new changes to the function. Please give your feedback.

parsoyaarihant commented:
@Shekharrajak , by making `check = False` [here](https://github.com/sympy/sympy/pull/11109/files/593f065e524f533a09106eb6015a8ace71266df3#r63389943) we get the correct answer for `solveset(x**2 + sqrt(x**2-5) >0,x,S.Reals)`. 

But, by changing `check= True` [this](https://github.com/parsoyaarihant/sympy/blob/solveset-bug-%2311064/sympy/solvers/tests/test_inequalities.py#L302) starts failing. In my local machine `solve(sqrt(x) + 1/root(x, 3) - 1,check=False)` give immediate result but when `solve(sqrt(x) + 1/root(x, 3) - 1,check=True)` it doesnt give any output.

arihantparsoya committed:
made changes to account for other equations

made changes to account for equation :
sqrt(x**2-5) < -x since sqrt(x**2 - 5) > 0 so -x > 0

arihantparsoya committed:
typo error

arihantparsoya committed:
added new test

arihantparsoya committed:
removed unnecessary lines

parsoyaarihant commented:
@Shekharrajak , can you review this now?

Shekharrajak commented:
```
In [1]: eq = 2 + sqrt(x - 4)

In [2]: solveset(eq > 0, x, S.Reals)
Out[2]: (-∞, 8) ∪ (8, ∞)

```
Not correct, right ?
Please don't skip the testcases.

arihantparsoya committed:
Made corrections in  test_slow_general_univariate()

parsoyaarihant commented:
@Shekharrajak , I was thinking that `check=False` is going to remain that way because of ` test_slow_general_univariate() ` in test_inequalities.py hence I deleted the test for `eq = 2 + sqrt(x - 4)` . But, I re-evaluated ` test_slow_general_univariate() ` and made the right correction to it. 

I am confident that the changes made to the test are correct since the minima of the function in test is greater that 1
![screen shot 2016-05-20 at 14 57 34](https://cloud.githubusercontent.com/assets/15258498/15424023/006bc05c-1e9e-11e6-8e7b-9c1514571734.png)
.

parsoyaarihant commented:
@aktech  , @Shekharrajak can your review this?

parsoyaarihant commented:
@Shekharrajak, can you review this?

smichr commented:
@parsoyaarihant , sorry that this hasn't gotten any attention for some time. Having forgot about this, I started some work along this line in #12471 . If you could give that a look I would appreciate it. I can then work with whatever is left over here to see that it gets completed, too.

parsoyaarihant commented:
@smichr , thanks for telling me. I will look into it soon.

czgdp1807 commented:
The issue has been fixed.
```python
>>> solveset(x+sqrt(x**2-5)>0,x,S.Reals)
[√5, ∞)
```
Closing this. Thanks for the contributions. Please let us know if there is a need to re-open this.
